RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,208 @@
+#define randlapack_comps_preconditioners_h
+
+#include "rl_orth.hh"
+#include "rl_rf.hh"
+#include "rl_blaspp.hh"
+#include "rl_lapackpp.hh"
+#include "rl_util.hh"
+
+#include <RandBLAS.hh>
+#include <math.h>
+#include <cstdint>
+#include <limits>
+#include <vector>
+
+#include <iostream>
+#include <cstdint>
+
+namespace RandLAPACK {
+
+/**
+ * Use sketching to produce data for a right-preconditioner for tall
+ * matrices of the form A_aug = [A; sqrt(mu) * I]. This is useful for ...
+ * 
+ *      Overdetermined and underdetermined least squares
+ *      ------------------------------------------------
+ *      It makes A_pre = A_aug*M well-conditioned, so that
+ *      min{||A_pre * z - b ||} and min{ ||y|| s.t. A_pre' y = c }
+ *      can easily be solved by unpreconditioned iterative methods.
+ * 
+ *      Solving linear systems with G = A'A + mu*I
+ *      ------------------------------------------
+ *      It makes M'GM well-conditioned, so that M'GM z = M'h
+ *      can easily be solved by CG without preconditioning.
+ * 
+ * The input matrix A must be contiguous; no "LDA" is allowed.
+ * 
+ * On exit, the preconditioner with mu=0 would be defined by
+ * taking the columns of V_sk and scaling them by 1/sigma_sk.
+ * We do not perform this scaling ourselves, since other values
+ * of mu might be intended.
+ * 
+ * @param[in] layout_A
+ *      Either blas::Layout::RowMajor or blas::Layout::ColMajor.
+ * @param[in] m
+ *      The number of rows in A
+ * @param[in] n
+ *      The number of columns in A (and rows in M)
+ * @param[in] d
+ *      The number of rows in the sketched data matrix; must be >= n.
+ * @param[in] k
+ *      The number of nonzeros per column in the sketching operator
+ *      A common value is k=8 when n is on the order of a couple thousand.
+ * @param[in] A
+ *      A buffer for an m \times n matrix A. Interpreted in 
+ *      "layout_A" order.
+ * @param[out] V_sk
+ *      We require that V_sk.size() >= d*n. Letting "r" denote the return
+ *      value of this function, on exit the orthogonal factor of the preconditioner
+ *      is identified with the first n*r entries in V_sk.data(),
+ *      read in column-major format.
+ * @param[in] sigma_sk
+ *      We require sigma_sk.size() >= n.
+ * @param[in] threads
+ *      The number of OpenMP threads used during sketching.
+ * @param[in] seed_key
+ *      The key provided to an underlying counter-based random number generator.
+ * @param[inout] seed_ctr
+ *      A 32-bit counter provided to the underlying counter-based random number
+ *      generator. This value is incremented by m*k on-exit.
+ * @returns
+ *      void
+ */
+template <typename T>
+void rpc_data_svd_saso(
+    blas::Layout layout_A,
+    int64_t m, // number of rows in A
+    int64_t n, // number of columns in A
+    int64_t d, // number of rows in sketch of A
+    int64_t k, // number of nonzeros in each column of the sketching operator
+    std::vector<T>& A, // buffer of size m*n.
```

Comment:
I can't use an std::vector API without a lot of copies. Is it possible to have a `T *` or `hamr::buffer<T>` API?

Reply:
Sure thing. This function doesn't benefit from memory tricks with std::vector anyway.